module/authentication/src/Infrastructure/Middleware/AuthenticationMiddleware.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/**
+ * Copyright © Bold Brand Commerce Sp. z o.o. All rights reserved.
+ * See LICENSE.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\Authentication\Infrastructure\Middleware;
+
+use Ergonode\Account\Application\Security\Security;
+use Ergonode\Account\Domain\Entity\User;
+use Ergonode\Account\Domain\Repository\UserRepositoryInterface;
+use Ergonode\Authentication\Infrastructure\Stamp\UserStamp;
+use Lexik\Bundle\JWTAuthenticationBundle\Security\Authentication\Token\JWTUserToken;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+use Symfony\Component\Messenger\Middleware\StackInterface;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+class AuthenticationMiddleware implements MiddlewareInterface
+{
+
+    private TokenStorageInterface $tokenStorage;
+
+    private UserRepositoryInterface $userRepository;
+
+    private Security $security;
+
+    public function __construct(
+        TokenStorageInterface $tokenStorage,
+        UserRepositoryInterface $userRepository,
+        Security $security
+    ) {
+        $this->tokenStorage = $tokenStorage;
+        $this->userRepository = $userRepository;
+        $this->security = $security;
+    }
+
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (!$envelope->last(ReceivedStamp::class)) {
+            /** @var User $user */
+            $user = $this->security->getUser();
+
+            if ($user) {
+                $envelope = $envelope->with(new UserStamp($user->getId()));
+            }
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        if ($envelope->last(ReceivedStamp::class)) {
+
+            $this->tokenStorage->setToken(null);
```

Comment:
I see here that we are asking for the user each time, this will be an extra load for each message which will affect their handling time. 
It would be possible to check if the logged  user is the same as the one we want to login and only if it is different, log new one in or logout him.

Reply:
For ReceivedStamp we should not have any user logged in BUT I'd actually check what Rafał is suggesting AND in other cases log out the logged-in user and add a log warning message for it.